### ISSUE TITLE
[Feature Request] Define a sweep in config.yaml file rather than command line

### ISSUE THREAD

**Issue by tomsercu** [facebookresearch/hydra#1376]
# 🚀 Feature Request

Currently a sweep can only be defined on the command line. Could it be supported through the config files as well?

## Motivation / pitch
Situations where this particularly makes sense: `python train.py seed="range(0,10)" xval_fold="range(0,5)" actual_hyperparam_I_want_to_change=1,2,3`
-- if for any new hyperparameter combination we always want to run over those several seeds and cross-validation folds.

For those cases I'd like to be able to have those sweeps in `config.yaml`, to declutter the command line.

Couldn't find prior discussions about this, but somewhat related is #394 

cc @liujas000

**Comment by jieru-hu** [comment]
Defining a sweep in config file is supported in all Hydra sweepers([example](https://hydra.cc/docs/next/plugins/ax_sweeper)) except for the BasicSweeper (the one you are using)

@omry does it make sense for us to add this support to the BasicSweeper?

**Comment by omry** [comment]
I think this can be added as a part of the change to standardize the configuration of the search space of HPO sweepers.

We have some work in that direction, but it's stalled. (cc @shagunsodhani).
I can't find a corresponding issue for this feature request.

To set expectations:
None of this is likely to happen in time for Hydra 1.1.

**Comment by zaccharieramzi** [comment]
I am very much interested in this feature, that's why I had opened a duplicate #1948 .

Reposting here an additional request that I had regarding this feature: do you think it would be possible to make it work with [this type of sweep](https://github.com/facebookresearch/hydra/issues/1258#issuecomment-754341537)? i.e. `python my_app.py -m '+top={a:10,b:20},{a:100,b:200}'`

**Comment by Jasha10** [comment]
> ... do you think it would be possible to make it work with ... `python my_app.py -m '+top={a:10,b:20},{a:100,b:200}'`


Modifying the [example] from PR #1886, I've had success with the following `config.yaml`:
```yaml
defaults:
  - db: ???
  - _self_

top: ???

hydra:
  sweeper:
    params:
      db: glob(*)
      db.timeout: 5,10
      top: "{a:10,b:20},{a:100,b:200}"
```
These settings for `hydra.sweeper.params` are equivalent to the following command-line overrides:
```text
python my_app.py -m 'top={a:10,b:20},{a:100,b:200}'  # note: no plus symbol in front of "top"
```

**Comment by zaccharieramzi** [comment]
Amazing, sorry I didn't try to test it earlier.
Thanks for your reponsiveness.
I think this feature will greatly improve my setup.

**Comment by Jasha10** [comment]
> I think this feature will greatly improve my setup.

Glad to hear it 😁

**Comment by zaccharieramzi** [comment]
@Jasha10 I have experimented a bit now, and I would like to know how you would go about trying to put the sweep part in a different configuration file (separate from the one defining the defaults and say the basic configuration parameters)?

I tried the following:
```
.
├── conf
│   ├── config.yaml
│   └── hydra
│       └── sweeper
│           └── params
│               └── grid_search.yaml
└── my_app.py
```

Where the content of `grid_search.yaml` is:
```yaml
compile:
  lr: 1e-2,1e-3,1e-4
```

and the content of `config.yaml` is:
```yaml
compile: ???
```

Running `python my_app.py -m hydra.sweeper.params=grid_search`, gives the following error:
```
Error merging override hydra.sweeper.params=grid_search
Cannot assign str to Dict[str, str]
    full_key: hydra.sweeper.params
    object_type=BasicSweeperConf

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

**Comment by Jasha10** [comment]
Thank you for testing the limits of this new feature, @zaccharieramzi.

First let me say that you will need `python -m my_app.py hydra/sweeper/params=grid_search` (with slashes) instead of `python -m my_app.py hydra.sweeper.params=grid_search` (with periods). I've just added an [update](https://github.com/facebookresearch/hydra/commit/bf5a767b6a34d1721a110563a8880cbeaefb4269) to the docs with examples.

Here are two methods to achieve what you're going for, one of which depends on a PR that I have just opened and which is not yet merged into the main branch.

### Method one: using an `experiment` directory:

This method works currently with the main branch of hydra.

```text
$ tree
.
├── conf
│   ├── config.yaml
│   └── experiment
│       └── grid_search.yaml
└── my_app.py

2 directories, 3 files
```
#### `conf/experiment/grid_search.yaml`
```yaml
# @package _global_
hydra:
  sweeper:
    params:
      +compile.lr: 1e-2,1e-3,1e-4
```
This is using an [@package directive](https://hydra.cc/docs/next/advanced/overriding_packages/#overriding-the-package-via-the-package-directive).
#### At the command line:
```text
$ python my_app.py -m +experiment=grid_search
[2022-01-12 20:23:38,705][HYDRA] Launching 3 jobs locally
[2022-01-12 20:23:38,705][HYDRA]        #0 : compile.lr=0.01 +experiment=grid_search
{'compile': {'lr': 0.01}}
[2022-01-12 20:23:38,790][HYDRA]        #1 : compile.lr=0.001 +experiment=grid_search
{'compile': {'lr': 0.001}}
[2022-01-12 20:23:38,882][HYDRA]        #2 : compile.lr=0.0001 +experiment=grid_search
{'compile': {'lr': 0.0001}}
```

### Method two: selecting an option for `'hydra/sweeper/params'`

This method depends on PR #1956 and is not yet enabled on the `main` Hydra branch.
```text
$ tree
.
├── conf
│   ├── config.yaml
│   └── hydra
│       └── sweeper
│           └── params
│               └── grid_search.yaml
└── my_app.py
```
#### `conf/hydra/sweeper/params/grid_search.yaml`
```yaml
+compile.lr: 1e-2,1e-3,1e-4
```
#### At the command line:
```text
$ python my_app.py -m hydra/sweeper/params=grid_search hydra.job.chdir=true
[2022-01-12 20:34:33,921][HYDRA] Launching 3 jobs locally
[2022-01-12 20:34:33,921][HYDRA]        #0 : compile.lr=0.01
{'compile': {'lr': 0.01}}
[2022-01-12 20:34:34,006][HYDRA]        #1 : compile.lr=0.001
{'compile': {'lr': 0.001}}
[2022-01-12 20:34:34,101][HYDRA]        #2 : compile.lr=0.0001
{'compile': {'lr': 0.0001}}
```

**Comment by zaccharieramzi** [comment]
@Jasha10 Thanks for this reply!

I personally like the first option better, even with the potential copy-pasting I will have to do for different grid searches, because it allows me to have an `experiment` directory, which is very informative.

It could be nice to think about having a shortcut for:
```yaml
# @package _global_
hydra:
  sweeper:
    params:
```
but certainly not for the next release.

I also tried the following syntax:
```yaml
# @package _global_
hydra:
  sweeper:
    params:
      +compile: "{lr:1e-2,wd:1e-4},{lr:1e-3,wd:1e-5}"
```
and it worked perfectly as expected.

Can't wait for the next release, and actually I will probably use it right away ahah!

**Comment by GiorgioBarnabo** [comment]
Hi @Jasha10,

thank you for the explaination. Unfortunately, neither option is working in my case.

When trying method one, I get this error:

![Schermata 2022-03-16 alle 14 56 15](https://user-images.githubusercontent.com/38131830/158606150-940eb95d-699f-4fc9-ad9b-d94b0db178c7.png)

While for method two, I get:

![Schermata 2022-03-16 alle 14 57 05](https://user-images.githubusercontent.com/38131830/158606298-910fc330-0d27-4797-8e8d-149f13618770.png)

Did something change? Maybe I am making mistakes in the config file? I kind of tried all possible combinations. 

This is the python script with my_app():


![Schermata 2022-03-16 alle 14 59 30](https://user-images.githubusercontent.com/38131830/158606907-336f7036-eef9-42f2-9a06-db2c2cd9cc8d.png)

Thank you in advance. Best,

Giorgio

**Comment by Jasha10** [comment]
@GiorgioBarnabo for method 1 to work, you will need to install a recent development version of Hydra (e.g. `pip install --upgrade hydra-core==1.2.0.dev2`) or install from the `main` branch here on github.

Method 2 is currently not working (as we have closed PR #1956).
Method 2 should work again once PR #2056 is finalized.
